File: test-suite/src/test/java/io/micronaut/kafka/docs/consumer/batch/BookListener.java
package io.micronaut.kafka.docs.consumer.batch;

// tag::imports[]
import io.micronaut.configuration.kafka.annotation.KafkaListener;
import io.micronaut.configuration.kafka.annotation.Topic;
import io.micronaut.context.annotation.Requires;
import org.apache.kafka.clients.consumer.Consumer;
import org.apache.kafka.clients.consumer.ConsumerRecord;
import org.apache.kafka.clients.consumer.OffsetAndMetadata;
import org.apache.kafka.common.TopicPartition;
import org.slf4j.Logger;
import reactor.core.publisher.Flux;

import java.util.Collections;
import java.util.List;

import static org.slf4j.LoggerFactory.getLogger;
// end::imports[]

@Requires(property = "spec.name", value = "BatchBookListenerTest")
// tag::clazz[]
@KafkaListener(batch = true) // <1>
public class BookListener {
// end::clazz[]
    private static final Logger LOG = getLogger(BookListener.class);

    // tag::method[]
    @Topic("all-the-books")
    public void receiveList(List<Book> books) { // <2>
        for (Book book : books) {
            LOG.info("Got Book = {}", book.title()); // <3>
        }
    }
    // end::method[]

    // tag::reactive[]
    @Topic("all-the-books")
    public Flux<Book> receiveFlux(Flux<Book> books) {
        return books.doOnNext(book ->
            LOG.info("Got Book = {}", book.title())
        );
    }
    // end::reactive[]

    // tag::manual[]
    @Topic("all-the-books")
    public void receive(List<ConsumerRecord<String, Book>> records, Consumer kafkaConsumer) { // <1>

        for (int i = 0; i < records.size(); i++) {
            ConsumerRecord<String, Book> record = records.get(i); // <2>

            // process the book
            Book book = record.value();

            // commit offsets
            String topic = record.topic();
            int partition = record.partition();
            long offset = record.offset(); // <3>

            kafkaConsumer.commitSync(Collections.singletonMap( // <4>
                    new TopicPartition(topic, partition),
                    new OffsetAndMetadata(offset + 1, "my metadata")
            ));

        }
    }
    // end::manual[]
}

File: test-suite-python/src/test/python/micronaut/kafka/docs/consumer/batch/BookListener.py
import logging

# tag::imports[]
from micronaut.configuration.kafka.annotation import KafkaListener, Topic
from micronaut.context.annotation import Requires
from reactor.core.publisher import Flux

from .Book import Book
# end::imports[]

LOG = logging.getLogger(__name__)


@Requires(property="spec.name", value="BatchBookListenerTest")
# tag::clazz[]
@KafkaListener(batch=True)  # <1>
class BookListener:
# end::clazz[]

    # tag::method[]
    @Topic("all-the-books")
    def receive_list(self, books: list[Book]) -> None:  # <2>
        for book in books:
            LOG.info("Got Book = %s", book.title)  # <3>
    # end::method[]

    # tag::reactive[]
    @Topic("all-the-books")
    def receive_flux(self, books: Flux[Book]) -> Flux[Book]:
        return books.doOnNext(lambda book: LOG.info("Got Book = %s", book.title))
    # end::reactive[]
#tag::endclazz[]
#end::endclazz[]
